pkg/services/ngalert/database.go

```diff
@@ -119,3 +122,65 @@ func (ng *AlertNG) getAlertDefinitions(cmd *listAlertDefinitionsCommand) error {
 		return nil
 	})
 }
+
+// saveAlertDefinition is a handler for saving a new alert definition.
+func (ng *AlertNG) saveAlertInstance(cmd *saveAlertInstanceCommand) error {
```

Comment:
I know this is early wip, but long term I would want more isolation of database access (ie in its own repository abstraction), to help separate concerns, aid testability etc. 

The main AlertNG should be just the top level service with very knowledge of implementation details.

Reply:
@torkelo  Yes, I think once we have the initial version of this and the scheduler merged, we should take a beat and figure out a good pkg layout under ngalert/.

As an aside, the structure under pkg/expr is also not making a sense, so that needs similar thought / attention.

Reply:
ok sounds good. 

I would not wait too long to think about the abstractions we need and different responsibilities & interfaces and dependencies between components

Reply:
> I would not wait too long to think about the abstractions we need and different responsibilities & interfaces and dependencies between components

Made an issue https://github.com/grafana/grafana/issues/29386 so we can start to figure this out and write down ideas so we are ready when there is a good point to rename, move stuff around, create interfaces.